### ISSUE TITLE
Union types in generics do not correctly check types.

### ISSUE THREAD

**Issue by jespertheend** [microsoft/TypeScript#46552]
# Bug Report

### 🔎 Search Terms

union generic implicit

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generic

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=13&ssc=1&pln=14&pc=1#code/JYOwLgpgTgZghgYwgAgCoHcD2BlOBbCVATwAcIBnAcQhGmAQB5UA+ZAbwChlk4AuNANxdkAI36ohAXw4cwpFBhz5C88gEFkAXjRZcBYmSo06jAHIBXPCOisAPjqX7V1WlHoNsYNyADmzIXJkDnoqhgBCWsHKBhQuJgwWVtDI9p7efkIcAPRZyAACYOQAtBAAHmQIYCVQUJhQHAiYIORgyKXI4rrRqhranNx8yACMADTCYsgARNC1UACEkyPIOcgMRUVoABbA5MgzdeRLcLtlFZAAJhzSDU0tyEQdUU7hkf08-KPj-NM1dQtLKzWG1Q21250wFGQIEwrR8wAAbig4CA9r8oEtrAg4OZyCgAAYiPHITbHZBgTYoQL4tKgHwpZCJaxQPFXIA)

### 💻 Code

```ts
interface TwoSameTypes<T> {
  a: T;
  b: T;
}

// @ts-expect-error
const x : TwoSameTypes<Number | String> = {
  a: 1,
  b: "string", // <-- This does not give an error, because `b` has the type `String | Number`
}
```

### 🙁 Actual behavior

No error happens

### 🙂 Expected behavior

I would expect an error for the value of `x` because both `a` and `b` are expected to have the same type, but they don't in this case.

For the record, and as can be seen from the playground link, when doing `TwoSameTypes<Number> | TwoSameTypes<String>` instead of `TwoSameTypes<Number | String>` we do end up with the desired behaviour. But I think these two different ways of defining are equivalent? So it seems like the latter way of doing it should work as well.

Furthermore, [here's demo of a use case](https://www.typescriptlang.org/play?ts=4.4.4#code/JYOwLgpgTgZghgYwgAgGpwDYFcIFk4AOyA3gFACQIWAtgFzIByNARtANwUDOYU9Ayj1ABzDgF9SpUJFiIUAMQD2CgDwAVZBAAekEABNOyANYQAngphpMOfAQB8JCmBMEI9VR3IA3TPXTY8hADaqgC6YhJOLsi4EGAAFgq6AILIALzIiioARFTUWfYAPhlKylncUPkckSgx8YkAQmnF2blZyEVlPJWk1dGxCboAwk2ZysZmFn7WhLYcpAD088gAAmCcALRaLghgm1BQClCkCAog3MiayPS1AynpZOTV9Dk0WQA0FN4Yz9AHUACE72Qi2QynW62QqjiwAMQmAnggBjgIA0+0Ob2QcAMWwgOwgulI4gWS1WGxxOz2f2Op3OJiufTquka90czlcyBeeQ+Xh8HN+h0BGJBYIhUJhyF0CkRyBACjAyDhCMxKP5UAxrAQcCwnBQAAMvrrkHEsch4iherqBFBhO1GCxoLrCRIQaTNpptrtVdSzvKAF70m6JYYsx5s56tblfH5ogFA4XgyHQpEYTgKCVSzggADk8sVKGRqL+yBgh1NcRQnDg1BQUAgWNOTqAA) for this behaviour. I would like to use `keyof ValueMap` as the generic type, so I don't have to type every possible enum twice.

Apologies in advance if this is a duplicate, I wasn't able to find the right keywords to describe this behaviour.
